pkg/auth/oauth/registry/tokentimeouts.go

```diff
@@ -0,0 +1,219 @@
+package registry
+
+import (
+	"errors"
+	"math"
+	"time"
+
+	"github.com/golang/glog"
+	"github.com/google/btree"
+
+	kerrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"github.com/openshift/origin/pkg/oauth/apis/oauth"
+	oauthclient "github.com/openshift/origin/pkg/oauth/generated/internalclientset/typed/oauth/internalversion"
+)
+
+var ErrTimedout = errors.New("Token timed out")
+
+type tokenData struct {
+	name    string
+	client  string
+	seen    time.Time
+	timeout time.Time
+}
+
+type tokenDataRef struct {
+	name    string
+	timeout time.Time
+}
+
+func (a *tokenDataRef) Less(than btree.Item) bool {
+
+	if a.timeout.Equal(than.(*tokenDataRef).timeout) {
+		return a.name < than.(*tokenDataRef).name
+	}
+	return a.timeout.Before(than.(*tokenDataRef).timeout)
+}
+
+type TokenTimeoutUpdater struct {
+	oauthClient    oauthclient.OAuthClientInterface
+	clientTimeouts map[string]int32
+	tokens         oauthclient.OAuthAccessTokenInterface
+	tokenChannel   chan tokenData
+	data           map[string]tokenData
+	tree           *btree.BTree
```

Comment:
Doesn't a heap give you the properties you need?

Reply:
No, not really,  a heap does not allow me to see the value before deciding if I want to get it, and it also does not allow you to replace elements map + tree give me those properties.

Reply:
Was hoping there was something in the std lib.  Oh wells.

Reply:
Couldn't find anything, glad to avoid custom code if you find anything workable.